### PR TITLE
[MIG 10.0] product_profile  & product_profile_example

### PR DESCRIPTION
**Pull request by EBII**
product_profile  : generic module
product_profile_example: module for test 

--test-enable --log-level=test -u product_profile_example -d

### CODE REVIEW

File: product_profile/i18n/de.po [hunk 1]
```diff
@@ -4,14 +4,13 @@
 #
```
Comment (sebastienbeau) [inline product_profile/i18n/de.po]: No need to modify the pot file

File: product_profile/product.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
+# coding: utf-8
```
Comment (sebastienbeau) [inline product_profile/product.py]: You should use git mv, This file is twice here

**Review by sebastienbeau — CHANGES_REQUESTED** [review]
You last commit is not correct

**Comment by EBII** [comment]
ping @sebastienbeau

File: product_profile/README.rst [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+Profiles are also defined as search filter and group.
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/product-attribute/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/product-attribute/issues/new?body=module:%20product_profile%0Aversion:%201.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
```
Comment (mourad-ehm) [inline product_profile/README.rst]: Update the Readme.rst to the last version. (short bug tracker)

File: product_profile/models/product.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    # These fields must not be synchronized between product.profile
+    # and product.template/product
+    return models.MAGIC_COLUMNS + [
+        'name', 'explanation', 'sequence',
+        'display_name', '__last_update']
+
+
+class ProductProfile(models.Model):
+    _name = 'product.profile'
+    _order = 'sequence'
```
Comment (mourad-ehm) [inline product_profile/models/product.py]: It is nice to have an order by name too: _order = 'sequence, name'

File: product_profile/models/product.py [hunk 5]
```diff
@@ -89,11 +79,8 @@ def write(self, vals):
             for rec in self:
                 products = self.env['product.product'].search(
                     [('profile_id', '=', rec.id)])
-                if products:
-                    _logger.info(
```
Comment (bealdav) [inline product_profile/models/product.py]: Why no more logger ? except that LGTM

**Comment by mourad-ehm** [comment]
Code review :+1:

File: product_profile/models/product.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                    self.check_useless_key_in_vals(new_vals, key)):
+                new_vals.pop(key)
+        # super call must be after check_useless_key_in_vals() call
+        # because we compare value before and after write
+        res = super(ProductProfile, self).write(new_vals)
+        if new_vals:
+            for rec in self:
+                products = self.env['product.product'].search(
+                    [('profile_id', '=', rec.id)])
+                _logger.info(" >>> %s Products updating after updated '%s' pro"
```
Comment (sebastienbeau) [inline product_profile/models/product.py]: Can you restore the if products: 
Thx

File: product_profile/models/product.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        return vals
+
+    @api.onchange('profile_id')
+    def _onchange_from_profile(self):
+        """ Update product fields with product.profile corresponding fields """
+        self.ensure_one()
+        if self.profile_id:
+            values = self._get_profile_data({'profile_id': self.profile_id.id})
+            for field, value in values.items():
+                if field in self._fields:
```
Comment (sebastienbeau) [inline product_profile/models/product.py]: Maybe this if is not needed? if needed add a comment

File: product_profile/models/product.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+    @api.model
+    def _get_profile_fields(self):
+        fields_to_exclude = set(get_profile_fields_to_exclude())
+        return [field for field in self.env['product.profile']._fields.keys()
+                if field not in fields_to_exclude]
+
+    @api.model
+    def _get_profile_data(self, values, filled_fields=None):
+        # Note for migration to v9
```
Comment (sebastienbeau) [inline product_profile/models/product.py]: Please rename the method _get_profile_data(self, values, filled_fields=None) => _get_vals_from_profile(self, values)

File: product_profile/models/product.py [hunk 9]
```diff
@@ -162,24 +160,24 @@ def _onchange_from_profile(self):
...
         if self.profile_id:
-            values = self._get_profile_data({'profile_id': self.profile_id.id})
+            values = self._get_vals_from_profile({'profile_id': self.profile_id.id})
             for field, value in values.items():
-                if field in self._fields:
-                    try:
-                        self[field] = value
-                    except Exception as e:
-                        raise UserError(format_except_message(e, field, self))
+               # if field in self._fields:
```
Comment (bealdav) [inline product_profile/models/product.py]: Why comment code ? Usefull or useless ?

**Review by bealdav — CHANGES_REQUESTED** [review]

**Comment by bealdav** [comment]
ping @sebastienbeau could you check last changes

Thanks

**Comment by mourad-ehm** [comment]
Hi @JesusZapata, @moylop260, 
I don't understand this Pylint errors on the readme file.
> ************* Module product_profile
product_profile/README.rst:7: [E7901(rst-syntax-error), ]  Enumerated list start value not ordinal-1: "2" (ordinal 2)
product_profile/README.rst:7: [E7901(rst-syntax-error), ]  Enumerated list start value not ordinal-1: "3" (ordinal 3)
product_profile/README.rst:7: [E7901(rst-syntax-error), ]  Enumerated list start value not ordinal-1: "4" (ordinal 4)

**Comment by JesusZapata** [comment]
@mourad-ehm 
The trouble is related with the Enumerated Lists 

The content within the numbered list must be aligned

For example : 

```rst
 1.  Create your own profile here: Sales > Configuration > Product > Product Profiles

     .. figure:: static/img/list.png                                              
          :alt: profile list                                                       
          :width: 600 px                                                          
                                                                                   
  2.  To have more fields available to attach to this profile you must define      
      these fields in the model 'product.profile' in your own module               
      If the field name (and its type) is the same than those in 'product.template'
      then values of these will be populated automatically                        
      in 'product.template'                                                        
      Example of fields declaration in your own module:                            
                                                                                   
      .. code-block:: python                                                      
                                                                                  
          class ProductProfile(models.Model):                                     
              """ Require dependency on sale, purchase and point_of_sale modules  
              """                                                                 
                                                                                  
              _inherit = 'product.profile'                                        
                                                                                  
              def _get_types(self):                                               
                  return [('product', 'Stockable Product'),                       
                      ('consu', 'Consumable'),                                    
                      ('service', 'Service')]      
```
Note: Delete the `|` chararter between each enumerated list

**Comment by moylop260** [comment]
and the flake8 red is because:
`product_profile_example/tests/test_profile.py:85:1: W391 blank line at end of file`

You can fixed it deleting the following line: [product_profile_example/tests/test_profile.py#L85](https://github.com/EBII/product-attribute/blob/e275f606f7b1358692acb0927e3df4d8a1f5db2f/product_profile_example/tests/test_profile.py#L85)

**Review by sebastienbeau — APPROVED** [review]

**Review by PierrickBrun — APPROVED** [review]

**Review by moylop260 — APPROVED** [review]

**Pull request merged**